# CHASSY-Multi-Omics-Analyisis
This repository contains a collection of scripts for mining, analysing and visualizing several omics data layers:
- Transcriptomics
- Relative and absolute proteomics
- Metabolomics

This datasets were produced for three different fungal organisms:
- *S. cerevisiae*
- *K. marxianus*
- *Y. lipolytica*

Growing on different experimental conditions:

| Conditions | *S. cervisiae* CEN.PK 113 /-D |	*K. marxianus* CBS397 |	*Y. lipolytica* W29 
| ------------- |:-------------:|:-------------:|:-------------:|
| Reference |	30°C / pH 5.5 |	30°C / pH 5.5 |	25°C / pH 5.5 |
| High temperature |	36°C / pH 5.5 |	44°C / pH 5.5 |	32°C / pH 5.5 |
| Low pH |	30°C / pH 3.5 |	30°C / pH 3.5 |	25°C / pH 3.5 |
| Osmotic stress |	30°C / pH 5.5 / KCL [1M] |	36°C / pH 5.5 / KCL [1M] |	--- |


The proteomics datasets have been incorporated to enzymatically constrained GEMs for the three organisms, available at:

| Organism | Model ID |	URL |
| ------------- |:-------------:|:-------------:|
| *S. cervisiae* |	ecYeastGEM |	https://github.com/SysBioChalmers/GECKO |
| *K. marxianus* |	ecKmarxGEM |	https://github.com/SysBioChalmers/EnzymeConstrained-Kmarx-GEM |
| *Y. lipolytica* |	ecYaliGEM |	https://github.com/SysBioChalmers/EnzymeConstrained-Yali-GEM |

version of the current genome-scale metabolic model of the ascomycetous yeast _Kluyveromyces marxianus_, available at:
https://github.com/SysBioChalmers/KluyveromycesMarxianus-GEM

- KeyWords

**Repo Category:** Data Analysis; **Utilisation:** Multi-omics and multi organisms datasets analysis; **Field:** Metabolic engineering; **Omic Source:** Transcriptomics, Proteomics, Metabolomics; **Taxonomy:** *S. cervisiae* CEN.PK 113 /-D,	*K. marxianus* CBS397, *Y. lipolytica* W29  
Last update: 2018-05-13

- The model:


This repository is administered by [@IVANDOMENZAIN](https://github.com/IVANDOMENZAIN), Division of Systems and Synthetic Biology, Department of Biology and Biological Engineering, Chalmers University of Technology

## Installation
### Required Software
- MATLAB (version 2013b)
- R studio (version 1.0.136 or later)
### Dependencies - Recommended Software
### Installation Instructions
* Clone master branch from [SysBioChalmers GitHub](https://github.com/SysBioChalmers/EnzymeConstrained-Yali-GEM).

